File: Dynamic Programming/Coin Change/Coin Change.cpp
#include <bits/stdc++.h>
#include <iostream>

using namespace std;

class Solution {
public:
  int recur(vector<int> coin, int amount, int index, vector<vector<int>> &dp) {
    if (index < 0 or amount < 0) {
      return INT_MAX;
    }
    if (amount == 0) {
      return 0;
    }
    if (dp[index][amount] != -1) {
      return dp[index][amount];
    }
    if (coin[index] < amount) {
      int exclude = recur(coin, amount, index - 1, dp);
      int include = recur(coin, amount - coin[index], index, dp);
      dp[index][amount] = min(include, exclude);
    } else {
      dp[index][amount] = recur(coin, amount, index - 1, dp);
    }
    return dp[index][amount];
  }
  int coinChange(vector<int> &coins, int amount) {
    vector<vector<int>> dp(coins.size() + 1, vector<int>(amount + 1, -1));
    int res = recur(coins, amount, coins.size() - 1, dp);
    return res == INT_MAX ? -1 : res;
  }
};

int main() { return 0; }

File: Dynamic Programming/Coin Change/Coin Change.py
from typing import *


class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        def recur(dp: Dict, index: int, amount: int) -> int:
            if amount < 0 or index < 0:
                return float("inf")
            if amount == 0:
                return 0
            if (index, amount) in dp:
                return dp[(index, amount)]

            if amount > 0:
                dp[(index, amount)] = min(
                    recur(dp, index - 1, amount),
                    recur(dp, index, amount - coins[index - 1]) + 1,
                )

            else:
                dp[(index, amount)] = recur(dp, index - 1, amount)
            return dp[(index, amount)]

        dp = {}
        res = recur(dp, len(coins) - 1, amount)
        return -1 if res == float("inf") else res
